File: BinaryExponentiation/ComputeProductOf2VeryLargeNumbersModuloM/Solution.py
# Function to multiply two very large numbers
def multiply(A, B, mod):
    ans = 0
    while B:
        if B & 1:
            ans = (ans + A) % mod
        A = (A + A) % mod
        B >>= 1
    return ans

# Main function
if __name__ == "__main__":
    mod = 10**9 + 7
    A = int(1e9)
    B = int(1e9)
    print(multiply(A, B, mod))

File: BinaryExponentiation/ComputeProductOf2VeryLargeNumbersModuloM/Solution.js
const mod = BigInt(1e9 + 7); // BigInt constant for the modulo operation

// Function to multiply two very large numbers
function multiply(A, B) {
    let ans = BigInt(0); // Initialize ans as BigInt
    while (B) {
        if (B & BigInt(1)) { // Bitwise AND operation, checking if the least significant bit is set
            ans = (ans + A) % mod; // BigInt arithmetic
        }
        A = (A + A) % mod; // BigInt arithmetic
        B >>= BigInt(1); // Bitwise right shift by 1 position
    }
    return ans;
}

// Main function
function main() {
    const A = BigInt(1e9); // BigInt constant for A
    const B = BigInt(1e9); // BigInt constant for B
    console.log(multiply(A, B).toString()); // Output the result as string
}

// Call the main function
main();
